src/libopensc/pkcs15-skeid.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * PKCS15 emulation layer for Slovak eID card
+ *
+ * Copyright (C) 2022 Juraj Šarinay <juraj@sarinay.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ *
+ * based on the PKCS15 emulation layer for EstEID card by Martin Paljak
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "common/compat_strlcpy.h"
+
+#include "internal.h"
+#include "log.h"
+#include "pkcs15.h"
+
+static const struct sc_aid skeid_aid_qes = {{0xE8, 0x28, 0xBD, 0x08, 0x0F, 0xA0, 0x00, 0x00, 0x08, 0x51, 0x00, 0x00, 0x11}, 13};
+
+static int sc_pkcs15emu_skeid_init(sc_pkcs15_card_t * p15card)
+{
+	int r;
+	int i;
+	int sn_len;
+	char *buf;
+
+	set_string(&p15card->tokeninfo->label, "eID karta");
+	set_string(&p15card->tokeninfo->manufacturer_id, "Ministerstvo vnutra SR");
+
+	sn_len = p15card->card->serialnr.len;
+	if (sn_len > 0) {
+		buf = malloc(2 * sn_len + 1);
+		if (!buf) return SC_ERROR_OUT_OF_MEMORY;
+		sc_bin_to_hex(p15card->card->serialnr.value, sn_len, buf,
+			2 * sn_len + 1, 0);
+		p15card->tokeninfo->serial_number = buf;
+	}
+
+	p15card->tokeninfo->flags = SC_PKCS15_TOKEN_PRN_GENERATION | SC_PKCS15_TOKEN_READONLY;
+
+	for (i = 0; i < 3; i++) {
+		const char *skeid_cert_names[3] = {
+			"Kvalifikovany certifikat",
```

Comment:
Not related to this PR, but it would be lovely to get some kind of i18n in OpenSC... A reminder-issue, not likely someone will work on this tho :)

Reply:
Nobody outside of EU/EIDAS bureaucrats and related techies call what is in human speak a "signature certificate" (maybe even a signature key) a "qualified certificate". EU bureaucrats also want us to use "qualified website certificates" but that is not happening for other technical/political/sane reasons. I'd recommend "precise enough but human readable labels" unless you claim that everyone in .sk speaks "legally valid terminology, also early in the morning before coffee" in which case I'll start using qualified website authentication certificates, also before morning coffee.

Reply:
Just wondering - is the "signing certificate" actually used for authentication? (technically a signing operation, but not qualified) ? If/how does the card support (online) authentication?

Reply:
I have started to support multiple languages in https://github.com/OpenSC/OpenSC/blob/master/src/ui/strings.h
Text is adjusted according to locale or via opensc.conf

Reply:
I dont think the language here is an issue as the eID cards are issued only to people with at least basic understanding of the Slovak language.

Reply:
@Jakuje, not true actually: the same chip is in the residence permits issues to foreigners, and also in the "alternative authenticator" cards issued to those who doesn’t have a residence permit.